File: inventory-framework-platform-bukkit/src/main/java/me/devnatan/inventoryframework/bukkit/pipeline/CancelledCloseInterceptor.java
package me.devnatan.inventoryframework.bukkit.pipeline;

import me.devnatan.inventoryframework.api.VirtualView;
import me.devnatan.inventoryframework.api.context.IFCloseContext;
import me.devnatan.inventoryframework.api.pipeline.PipelineContext;
import me.devnatan.inventoryframework.api.pipeline.PipelineInterceptor;
import me.devnatan.inventoryframework.bukkit.context.CloseContext;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

public class CancelledCloseInterceptor implements PipelineInterceptor<VirtualView> {

    @SuppressWarnings("ConstantValue")
    @Override
    public void intercept(PipelineContext<VirtualView> pipeline, VirtualView subject) {
        if (!(subject instanceof IFCloseContext)) return;

        final CloseContext context = (CloseContext) subject;
        if (!context.isCancelled()) return;

        final Player player = context.getPlayer();
        final ItemStack cursor = player.getItemOnCursor();

        context.getRoot().nextTick(() -> context.getViewer().open(context.getContainer()));

        // suppress cursor null check since it can be null in legacy versions
        if ((cursor != null) && cursor.getType() != Material.AIR) player.setItemOnCursor(cursor);
    }
}

File: inventory-framework-platform-minestom/src/main/kotlin/me/devnatan/inventoryframework/minestom/pipeline/CancelledCloseInterceptor.kt
package me.devnatan.inventoryframework.minestom.pipeline

import me.devnatan.inventoryframework.api.VirtualView
import me.devnatan.inventoryframework.api.pipeline.PipelineContext
import me.devnatan.inventoryframework.api.pipeline.PipelineInterceptor
import me.devnatan.inventoryframework.minestom.context.CloseContext

class CancelledCloseInterceptor : PipelineInterceptor<VirtualView> {
    override fun intercept(
        pipeline: PipelineContext<VirtualView>,
        subject: VirtualView,
    ) {
        if (subject !is CloseContext) return

        if (!subject.isCancelled) return

        subject.root.nextTick { subject.viewer.open(subject.container) }
    }
}
